modules/cli/src/main/java/org/apache/ignite/cli/core/repl/completer/HoconDynamicCompleter.java

```diff
@@ -0,0 +1,106 @@
+
+    private final Config config;
+
+    /** Stores all keys that exist in the given config. Strings are stored in format: "root.subkey.subsubkey". */
+    private final Set<String> leafs;
+
+    /** List of all possible completions. */
+    private final List<String> completions;
+
+    /** Words, after those the completer should have been activated. */
+    private final Set<String> activationPostfixes;
+
+    /** Default constructor that created an instance from given set of postfixes after those the completion is expected. */
+    public HoconDynamicCompleter(Set<String> activationPostfixes, Config config) {
+        this.activationPostfixes = activationPostfixes;
+        this.config = config;
+        this.leafs = config.entrySet().stream().map(Entry::getKey).collect(Collectors.toSet());
+        this.completions = this.compile();
+    }
+
+    private List<String> compile() {
+        ArrayList<String> result = new ArrayList<>();
+
+        String rootPrefix = "";
+        walkAndAdd(rootPrefix, config.root().keySet(), result);
+
+        return result;
+    }
+
+    /**
+     * Completes the given typed words with the config keys that a in the same level as the last typed words.
+     * <p/>
+     * Example: given typed words ["cluster", "config", "show", "--selector", "a"], The last word is "a", only root config values will be
+     * suggested to autocomplete: "aimem", "aipersist". If user hits "aimem" and types dot "." then only subkeys of "aimem." will be
+     * suggested: "aimem.pageSize", "aimem.regions".
+     */
+    @Override
+    public List<String> complete(String[] words) {
+        final String lastWord = findLastNotEmptyWord(words);
+
+        if (activationPostfixes.contains(lastWord)
+                // activation profile contains empty string and the last typed word is empty
+                || (activationPostfixes.contains("") && words[words.length - 1].isEmpty())) {
+            // roots
+            return completions.stream().filter(s -> s.split("\\.").length == 1).collect(Collectors.toList());
+        }
+
+        final int deepLevel = lastWord.endsWith(".")
+                ? lastWord.split("\\.").length + 1
+                : lastWord.split("\\.").length;
+
+        return completions.stream()
+                .filter(s -> s.startsWith(lastWord) && deepLevel == s.split("\\.").length)
+                .collect(Collectors.toList());
+    }
+
+    private String findLastNotEmptyWord(String[] words) {
+        for (int i = words.length - 1; i >= 0; i--) {
+            if (!words[i].isEmpty()) {
+                return words[i];
+            }
+        }
+        return "";
+    }
+
+    private void walkAndAdd(String keyPrefix, Set<String> keySet, List<String> result) {
+        keySet.forEach(key -> {
+            if (!leafs.contains(keyPrefix + key)) {
+                Set<String> nextKeySet = new HashSet<>(config.getConfig(keyPrefix + key).root().keySet());
```

Comment:
What's the point of wrapping the keySet in the new HashSet?

Reply:
You are right, there is no reason for that.